pkg/network/common/egressip.go

```diff
@@ -659,13 +933,68 @@ func (eit *EgressIPTracker) ReallocateEgressIPs() map[string][]string {
 		return allocation
 	}
 
-	// Compare the allocation to what we would have gotten if we started from scratch,
-	// to see if things have gotten too unbalanced. (In particular, if a node goes
-	// offline, gets emptied, and then comes back online, we want to move a bunch of
-	// egress IPs back onto that node.)
+	// Compare the allocation to what we would have gotten if we started from
+	// scratch, to see if things have gotten too unbalanced or if we can assign
+	// more egress IPs globally across the cluster.
 	fullReallocation, alreadyAllocated := eit.makeEmptyAllocation()
 	eit.allocateNewEgressIPs(fullReallocation, alreadyAllocated)
 
+	// Given a capacity constraint allocation problem: we need to consider cases
+	// where a full re-allocation enables more egress IPs to get assigned
+	// globally across the cluster w.r.t what was already assigned in the
+	// previous round of assignment. This would never be a case to consider when
+	// capacity is a moot point, because a full re-allocation should always
+	// assign as many egress IPs globally as the previous round of assignment
+	// (but might lack balance - but that is rectified below). An example of
+	// this is:
+
+	// Node 1 - capacity 2
+	// Node 2 - capacity 1
+
+	// Namespace 1 - [IP1]
+	// Namespace 2 - [IP3]
+
+	// An optimal allocation is:
+
+	// Node 1 - [IP1]
+	// Node 2 - [IP3]
+
+	// Now say the following changes:
+
+	// Namespace 1 - [IP1, IP2]
+
+	// An optimal allocation is now:
+
+	// Node 1 - [IP1, IP3]
+	// Node 2 - [IP2]
+
+	// We could however compute `allocation` above to be:
+
+	// Node 1 - [IP1]
+	// Node 2 - [IP3]
+
+	// Since that is also a fully valid solution based on the previous round of
+	// assignment. That however does not have as many global assignments as the
+	// optimal one described just above. Hence we need to always return directly
+	// if fullRelallocation has more global assignments than allocation. One
+	// sub-optimal solution to this problem would have been a final solution,
+	// such as:
+
+	// Node 1 - [IP2, IP3]
+	// Node 2 - [IP1]
+
+	// since that would have meant more moves (entropy) for the same final
+	// amount of global allocations. This will however never happen because
+	// allocateNewEgressIPs considers alreadyAllocated and will guarantee that
+	// we cover the case of allocating the maximum amount of egress IPs, in the
+	// least possible amount of moves.
+	if globalEgressIPCount(fullReallocation) > globalEgressIPCount(allocation) {
+		return fullReallocation
```

Comment:
I guess the question I have is: how do we chose which IPs to move or not? If we end up here: we know a full re-allocation will result in more assignments overall, though it might mean more changes. What do we decide is better then? I wasn't sure when coding this, and I'm still not. Do we prefer less changes at all costs, or more assignments? The example I have in the doc bloc kind of illustrates the dilemma: 
```
	// Node 1 - capacity 2
	// Node 2 - capacity 1

	// Namespace 1 - [IP1]
	// Namespace 2 - [IP3]

	// An optimal allocation is:

	// Node 1 - [IP1]
	// Node 2 - [IP3]

	// Now say the following changes:

	// Namespace 1 - [IP1, IP2]

	// An optimal allocation is now:

	// Node 1 - [IP1, IP3]
	// Node 2 - [IP2]
```

should we do this and gain one assignment at the expense of moving something which is already valid? Or just leave things the way they are? 

My reasoning was: let's always prefer more assignments, since removing / adding IPs to nodes _should_ always work, and if it doesn't, then that's the core problem.

Reply:
I guess the best thing might be to just to perform a cost-benefit analysis and compute the amount of changes required by going from `allocation` to `fullReallocation` and compare that to the amount of assignments won. If the assignments won differ by X something, accept the amount of changes required, else don't.

Reply:
> My reasoning was: let's always prefer more assignments, since removing / adding IPs to nodes should always work

But it's disruptive; any open TCP connections through the egress IP will get broken when the egress IP moves, and new connections may fail during the move.

> The example I have in the doc bloc kind of illustrates the dilemma

OK, but it's sort of artificial because the smallest possible change in that example is also pretty much the largest possible change. But if you have 10 egress nodes with capacity 10 each, and 9 of them have 10 egress IPs assigned and the last one has 9 egress IPs, and you add 1 more egress IP, and it can't be put onto the non-full node, we absolutely do not want to reassign _all 99 existing egress IPs_ to new nodes, which is something that might happen if we just blindly accept the full reallocation. We need to find a minimal(-ish) set of egress IPs that we can move around that will get a good result. I'm not sure what the best way to do that is... (the fact that we have to worry about capacity makes this harder than the existing cases).

If we don't have a good algorithm for figuring out a minimal change, then we should just fail to allocate the new egress IP in this case. (And maybe we can come back and try to improve things later...)

Reply:
> If we don't have a good algorithm for figuring out a minimal change, then we should just fail to allocate the new egress IP in this case

OK, let's just do that for now. I will remove this check completely and settle for `allocation` consistently